// ngx-datepicker-material/esm5/lib/daterangepicker.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MatButtonModule, MatIconModule } from '@angular/material';
import { DaterangepickerComponent } from './daterangepicker.component';
import { DaterangepickerDirective } from './daterangepicker.directive';
var NgxDaterangepickerMd = /** @class */ (function () {
    function NgxDaterangepickerMd() {
    }
    NgxDaterangepickerMd.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        DaterangepickerComponent,
                        DaterangepickerDirective
                    ],
                    imports: [
                        CommonModule,
                        FormsModule,
                        MatButtonModule,
                        MatIconModule,
                        ReactiveFormsModule
                    ],
                    providers: [],
                    exports: [
                        DaterangepickerComponent,
                        DaterangepickerDirective
                    ],
                    entryComponents: [
                        DaterangepickerComponent
                    ]
                },] }
    ];
    return NgxDaterangepickerMd;
}());
export { NgxDaterangepickerMd };
